core/sr-primitives/src/lib.rs

```diff
@@ -819,8 +884,41 @@ mod tests {
 	}
 
 	#[test]
-	fn saturating_mul() {
-		assert_eq!(super::Perbill::one() * std::u64::MAX, std::u64::MAX/1_000_000_000);
-		assert_eq!(super::Permill::from_percent(100) * std::u64::MAX, std::u64::MAX/1_000_000);
+	fn per_things_should_work() {
+		use super::{Perbill, Permill};
+		use primitive_types::U256;
+
+		per_thing_mul_upper_test!(u32, Perbill);
+		per_thing_mul_upper_test!(u64, Perbill);
+		per_thing_mul_upper_test!(u128, Perbill);
+
+		per_thing_mul_upper_test!(u32, Permill);
+		per_thing_mul_upper_test!(u64, Permill);
+		per_thing_mul_upper_test!(u128, Permill);
+	}
+
+	#[test]
+	#[should_panic]
+	fn per_things_operate_in_output_type() {
+		use super::Perbill;
+
+		assert_eq!(Perbill::one() * 255_u64, 255);
+		// panics
+		assert_ne!(Perbill::one() * 255_u8, 255);
```

Comment:
hmm yes I expect `As<u64>` to be a type that contains u64 which is false. maybe we should bound `Fomr<u32>` as well so it can't panic.

Reply:
maybe we shouldn't bound As but only `From<u32>` and `TryInto<u32>` actually it would work and never panic: we convert to u32 the remainder which is less that a billion.

Reply:
the thing is also that `From<u32>` is not implemented for simple arithmetic nor balance generic. And bounding a new trait to Currency::Balance is not super handy. Though we will want this at some point when we get rid of As

Reply:
or we could also bound some stuff like `Rem<u32, Output=N>`

Reply:
All of the above additional bounds though would have to be **enforced** to `Balance`, right? if so then I would say we should leave it as it is (test can also stay as a demonstration), as I don't see this as really _our concern_. 

These are utilities that we use and provide to substrate users as well. In our case, our balance types are actually big enough to work fine with this impl; And as for the users, this seems to be the _most generally safe_ approach. There will always be anomalies, but in this case, it will be a balance type of u16 and u8 which is not really realistic. 

What we could ultimately do to favor these group (or they can just do it themselves) is to create an actual `struct percent(u8)`. In this case, all u16 and u8 balances will also have sane arithmetic (I haven't done this but it looks to work).

Unless if there is an obvious simple way that we don't see, I see the current code as okay.

Reply:
Noting that if, for example, if your balance is u8, the granularity of a `per-thing` being more than 1/255 is actually pointless as there is no way to reflect it on the actual data type.

Reply:
not really it can be bound on module side by having an associated type `Balance: From<u32>` and write Curreny associated type like `Currency<Balance=Self::Balance>` but it creates a new associated type which is not beautiful.